### PR TITLE
Fix device manager to scan resources.Requests

### LINKED ISSUE #57276: Device manager only considers extended resource limits when allocating devices.
/kind bug
/sig node
/area hw-accelerators

**What happened**:
In `/pkg/kubelet/cm/deviceplugin/manager.go` devices are allocated from a function called `allocateContainerResources`. This function inspects only container resource limits, ignoring container resource requests. This is a usability problem because it is valid to omit limits for extended resources. If limit is omitted for a device resource, then the scheduler properly places the pod with respect to available node resources. However, the device manager never dispatches to the device plugin for that resource. The result is that the user application cannot access the requested device from inside the container.

**What you expected to happen**:
Following the documented behavior of extended resources, I expected that specifying a container resource request for a device would cause the device subsystem to properly configure access for my workload.

**How to reproduce it (as minimally and precisely as possible)**:
* Launch an alpha GKE cluster with GPU support enabled.
* Attempt to run any program that accesses the GPU from a container where the resource is specified in requests but omitted from limits.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.8+
- Cloud provider or hardware configuration: device plugins enabled, devices available.

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
**What this PR does / why we need it**:
This PR makes device manager to scan resources.Requests from the container spec. Currently
it scans resources.Limits. For extended resources, it is not mandatory for resources.Limits to be present in the container spec and if Limits are present, validation logic ensures that Limits will always be equal to Requests. 

Fixes #57276 

**Special notes for your reviewer**:

**Release note**:

```release-note
None
```
/sig node

/cc @ConnorDoyle @vishh @jiayingz @RenaudWasTaken @tengqm @resouer @mindprince

### CODE REVIEW

**Comment by RenaudWasTaken** [comment]
@vikaschoudhary16 the issue @ConnorDoyle opened was about the fact that we don't scan Requests but we still have to scan Limits to :)

**Comment by ConnorDoyle** [comment]
@RenaudWasTaken regarding

> we still have to scan Limits

Per the docs, requests default to limits if those are explicitly defined. For example, the scheduler only inspects requests but it is able to properly place pods that only specify limit.

So, it appears that requests are guaranteed to be a subset of limits. You can verify the behavior like this:

```
$ kubectl create -f - <<POD
apiVersion: v1
kind: Pod
metadata:
  name: er-test
spec:
  containers:
  - image: busybox
    name: er-test
    resources:
      requests:
        cpu: 50m
      limits:
        cpu: 100m
        example.com/foo: 1
POD
```

Then fetch the same pod resource from the API server with kubectl:

```
$ kubectl get pod er-test -o json | jq ".spec.containers[0].resources"
```

```json
{
  "limits": {
    "cpu": "100m",
    "example.com/foo": "1"
  },
  "requests": {
    "cpu": "50m",
    "example.com/foo": "1"
  }
}
```

In summary, I think it is safe to only inspect requests here and ignore limits. We will see any ER limit reflected in requests, and we are further guaranteed that ER limit is either undefined or equal to request.

**Comment by ConnorDoyle** [comment]
Thanks for posting a PR for this so quickly!

File: pkg/kubelet/cm/deviceplugin/manager.go [hunk 1]
```diff
@@ -550,7 +550,7 @@ func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Cont
 	podUID := string(pod.UID)
 	contName := container.Name
 	allocatedDevicesUpdated := false
-	for k, v := range container.Resources.Limits {
+	for k, v := range container.Resources.Requests {
```
Comment (ConnorDoyle) [inline pkg/kubelet/cm/deviceplugin/manager.go]: @vikaschoudhary16 it's might not be obvious why we can safely ignore Limits. Could we add a comment near here explaining the rationale?

**Comment by RenaudWasTaken** [comment]
@ConnorDoyle agreed but isn't the behavior undefined if someone sets requests and limits ? Should this error out ?

**Comment by ConnorDoyle** [comment]
You can set both request and limit for extended resources, but they fail validation if they are not equal.

File: pkg/kubelet/cm/deviceplugin/manager.go (hunk 1 above)
Comment (vikaschoudhary16) [reply pkg/kubelet/cm/deviceplugin/manager.go]: sure!

**Comment by ConnorDoyle** [comment]
/approve
Will add LGTM after the requested comments have been added.

**Comment by vikaschoudhary16** [comment]
@ConnorDoyle Done. PTAL!

File: pkg/kubelet/cm/deviceplugin/manager.go [hunk 2]
```diff
@@ -550,7 +550,10 @@ func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Cont
 	podUID := string(pod.UID)
 	contName := container.Name
 	allocatedDevicesUpdated := false
-	for k, v := range container.Resources.Limits {
+	// NOTE: Skipping the Resources.Lists is safe here because:
```
Comment (ConnorDoyle) [inline pkg/kubelet/cm/deviceplugin/manager.go]: s/Lists/Limits

**Review by ConnorDoyle — CHANGES_REQUESTED** [review]

File: pkg/kubelet/cm/deviceplugin/manager.go [hunk 3]
```diff
@@ -550,7 +550,10 @@ func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Cont
...
+	// 1. If container Spec mentions Limits only, implicitly Requests, equal to Limits, get added to Spec.
+	// 2. If container Spec mentions Limits, which are greater than or less than Requests, will fail at validation.
```
Comment (resouer) [inline pkg/kubelet/cm/deviceplugin/manager.go]: Sorry, I can't recall we have validation to verify 2) , `limits` can be greater than `requests` by design in scheduler ... or I misunderstood the comment here?

**Review by resouer — COMMENTED** [review]
I can confirm using `requests` here is safe and reasonable from scheduling aspect, so code lgtm, but comments may need update.

File: pkg/kubelet/cm/deviceplugin/manager.go [hunk 4]
```diff
@@ -550,7 +550,10 @@ func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Cont
 	podUID := string(pod.UID)
 	contName := container.Name
 	allocatedDevicesUpdated := false
-	for k, v := range container.Resources.Limits {
+	// NOTE: Skipping the Resources.Lists is safe here because:
+	// 1. If container Spec mentions Limits only, implicitly Requests, equal to Limits, get added to Spec.
```
Comment (resouer) [inline pkg/kubelet/cm/deviceplugin/manager.go]: The commas around `equal to Limits` confused me for a little bit. Not sure how do @ConnorDoyle feel? (guess you're native speak :P

**Comment by jiayingz** [comment]
@vikaschoudhary16 we may consider to cherry-pick this PR to 1.9.1.

File: pkg/kubelet/cm/deviceplugin/manager.go [hunk 5]
```diff
@@ -550,7 +550,10 @@ func (m *ManagerImpl) allocateContainerResources(pod *v1.Pod, container *v1.Cont
 	podUID := string(pod.UID)
 	contName := container.Name
 	allocatedDevicesUpdated := false
-	for k, v := range container.Resources.Limits {
+	// NOTE: Skipping the Resources.Limits is safe here because:
+	// 1. If container Spec mentions Limits only, implicitly Requests, equal to Limits, will get added to the Spec.
+	// 2. If container Spec mentions Limits, which are greater than or less than Requests, will fail at validation.
```
Comment (vishh) [inline pkg/kubelet/cm/deviceplugin/manager.go]: What if limits are not set? I think this change should validate if requests==limits prior to using requests.

**Comment by vishh** [comment]
I'd appreciate if API behavioral changes are soaked for a bit before being merged.

File: pkg/kubelet/cm/deviceplugin/manager.go (hunk 5 above)
Comment (vikaschoudhary16) [reply pkg/kubelet/cm/deviceplugin/manager.go]: @vishh if requests != limits for ERs, it will fail here on validation step:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L4380
And if limits are not set, then it should not be a problem because now with this change, requests are being scanned.

Or i missed your question?

Comment (resouer) [reply pkg/kubelet/cm/deviceplugin/manager.go]: @vishh Yeah, I raised the same question before, it is ensured by validation indeed, while I suggested to update this comment to make it more clear. 😆
